src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -506,17 +506,24 @@ pub fn prepare_init(cx: &mut Context, unit: &Unit) -> CargoResult<()> {
     let new2 = new1.clone();
 
     if fs::metadata(&new1).is_err() {
-        try!(fs::create_dir(&new1));
+        try!(fs::create_dir_all(&new1));
     }
     if fs::metadata(&new2).is_err() {
-        try!(fs::create_dir(&new2));
+        try!(fs::create_dir_all(&new2));
     }
     Ok(())
 }
 
 /// Return the (old, new) location for fingerprints for a package
 pub fn dir(cx: &Context, unit: &Unit) -> PathBuf {
-    cx.layout(unit).proxy().fingerprint(unit.pkg)
+    let dir = cx.layout(unit).proxy().fingerprint(unit.pkg);
+    if let Some(features) = cx.resolve.features(unit.pkg.package_id()) {
+        let mut feat_vec: Vec<String> = features.iter().cloned().collect();
+        feat_vec.sort();
+        dir.join(format!("feat-{}", feat_vec.join("-")))
+    } else {
+        dir
+    }
```

Comment:
This is correctly separating out the fingerprints, but not the executables.

The executables/libraries are correctly separated as long as they are a full dep, but when they are a path dep, I think this is not working right.